mantidproject/mantid
# Skip orphaned detectors in GenerateGroupingPowder

`GenerateGroupingPowder` grouped all detectors it could find in a workspace. This did not play well with `GroupDetectors` if the detectors were not attached to any spectrum. This PR fixes the problem by looping only over the detectors which are found in the input workspace's spectra.

**Report to:** [nobody].

**To test:**

Check that the call to `GroupDetectors` in the following script does not generate any errors/warnings in Mantid's logs:

```python
ws = CreateSampleWorkspace()
MaskDetectors(ws, '1,2,3')
ws = RemoveMaskedSpectra(ws)
filename = 'delete_me.xml'
GenerateGroupingPowder(ws, 0.2, filename, False)
grouped = GroupDetectors(ws, MapFile=filename)
```

Fixes #24560.

---

#### Reviewer ####

Please comment on the following ([full description](http://developer.mantidproject.org/IndividualTicketTesting/)):

##### Code Review #####

- Is the code of an acceptable quality?
- Does the code conform to the [coding standards](http://developer.mantidproject.org/Standards/)?
- Are the unit tests small and test the class in isolation?
- If there are changes in the release notes then do they describe the changes appropriately?

##### Functional Tests #####

- Do changes function as described? Add comments below that describe the tests performed?
- Do the changes handle unexpected situations, e.g. bad input?
- Has the relevant (user and developer) documentation been added/updated?

Does everything look good? Mark the review as **Approve**. A member of `@mantidproject/gatekeepers` will take care of it.
